File: leetcode/queue_stack/3221_Maximum_Array_Hopping_Score_II/codes/li_chao.c
#include <stdio.h>
#include <stdlib.h>
#include <limits.h>

typedef long long ll;

typedef struct { 
    ll m, b; 
} Line;

Line *st;

ll eval(Line ln, int x){ 
    return ln.m * x + ln.b; 
}

void add_line(int idx, int l, int r, Line nw){
    Line cur = st[idx];
    int mid = (l + r) >> 1;
    int left = eval(nw, l) > eval(cur, l);
    int m = eval(nw, mid) > eval(cur, mid);
    if(m){ Line tmp = st[idx]; st[idx] = nw; nw = tmp; cur = st[idx]; }
    if(l == r) return;
    if(left != m) add_line(idx<<1, l, mid, nw);
    else add_line(idx<<1|1, mid + 1, r, nw);
}
ll query_line(int idx, int l, int r, int x){
    ll res = eval(st[idx], x);
    if(l == r) return res;
    int mid = (l + r) >> 1;
    if(x <= mid){
        ll t = query_line(idx<<1, l, mid, x);
        return t > res ? t : res;
    } else {
        ll t = query_line(idx<<1|1, mid + 1, r, x);
        return t > res ? t : res;
    }
}
long long maxScore(int* nums, int numsSize) {
    if(numsSize == 0) return 0;
    int n = numsSize;
    int size = 4 * (n + 5);
    st = (Line*) malloc(size * sizeof(Line));
    for(int i = 0; i < size; ++i){ 
        st[i].m = 0;
        st[i].b = LLONG_MIN/4; 
    }

    ll *dp = (ll*) calloc(n, sizeof(ll));
    add_line(1, 0, n - 1, (Line){ - (ll)nums[n - 1], (ll)(n - 1) * nums[n - 1] + dp[n - 1] });
    for(int i = n - 2; i >= 0; --i){
        ll best = query_line(1, 0, n - 1, i);
        dp[i] = best;
        add_line(1, 0, n - 1, (Line){ - (ll)nums[i], (ll)i * nums[i] + dp[i] });
    }
    ll ans = dp[0];
    free(st);
    free(dp);
    return ans;
}

File: leetcode/queue_stack/3221_Maximum_Array_Hopping_Score_II/codes/li_chao.py
from typing import List

class LiChao:
    def __init__(self, l, r):
        self.l = l; 
        self.r = r
        self.line = None
        self.left = None
        self.right = None
    def eval(self, line, x):
        return line[0] * x + line[1]
    def add(self, new, l=None, r=None):
        if l is None:
            l = self.l; r = self.r
        if self.line is None:
            self.line = new
            return
        mid = (l + r) >> 1
        if self.eval(new, mid) > self.eval(self.line, mid):
            self.line, new = new, self.line
        if l == r:
            return
        if self.eval(new, l) > self.eval(self.line, l):
            if self.left is None:
                self.left = LiChao(l, mid)
            self.left.add(new, l, mid)
        elif self.eval(new, r) > self.eval(self.line, r):
            if self.right is None:
                self.right = LiChao(mid + 1, r)
            self.right.add(new, mid + 1, r)
    def query(self, x, l=None, r=None):
        if l is None:
            l = self.l; r = self.r
        if self.line is None:
            return -10**30
        res = self.eval(self.line, x)
        if l == r:
            return res
        mid = (l + r) >> 1
        if x <= mid:
            if self.left is None:
                return res
            return max(res, self.left.query(x, l, mid))
        else:
            if self.right is None:
                return res
            return max(res, self.right.query(x, mid + 1, r))

class Solution:
    def maxScore(self, nums: List[int]) -> int:
        n = len(nums)
        if n == 0:
            return 0
        root = LiChao(0, n - 1)
        dp = [0] * n
        root.add((-nums[n - 1], (n - 1) * nums[n - 1] + dp[n - 1]))
        for i in range(n-2, -1, -1):
            best = root.query(i)
            dp[i] = best
            root.add((-nums[i], i * nums[i] + dp[i]))
        return dp[0]
